--- xf86-video-intel-2.2.1/src/i830_display.c	2008-09-04 10:32:26.000000000 +0800
+++ xf86-video-intel-2.2.1/src/i830_display.c_change	2008-09-04 10:34:04.000000000 +0800
@@ -1196,6 +1196,14 @@ i830_crtc_mode_set(xf86CrtcPtr crtc, Dis
 	i830PrintPll("chosen", &clock);
     }
 
+    if (dpll & DPLL_VCO_ENABLE)
+    {
+        OUTREG(fp_reg, fp);
+	OUTREG(dpll_reg, dpll & ~DPLL_VCO_ENABLE);
+	POSTING_READ(dpll_reg);
+	usleep(150);
+    }
+
     /* The LVDS pin pair needs to be on before the DPLLs are enabled.
      * This is an exception to the general rule that mode_set doesn't turn
      * things on.
@@ -1204,14 +1212,6 @@ i830_crtc_mode_set(xf86CrtcPtr crtc, Dis
     {
 	CARD32 lvds = INREG(LVDS);
 
-	if (dpll & DPLL_VCO_ENABLE)
-	{
-	    OUTREG(fp_reg, fp);
-	    OUTREG(dpll_reg, dpll & ~DPLL_VCO_ENABLE);
-	    POSTING_READ(dpll_reg);
-	    usleep(150);
-	}
-
 	lvds |= LVDS_PORT_EN | LVDS_A0A2_CLKA_POWER_UP | LVDS_PIPEB_SELECT;
 	/* Set the B0-B3 data pairs corresponding to whether we're going to
 	 * set the DPLLs for dual-channel mode or not.
