diff -up xf86-video-i810-1.6.5/src/i830_driver.c.jx xf86-video-i810-1.6.5/src/i830_driver.c
--- xf86-video-i810-1.6.5/src/i830_driver.c.jx	2009-05-14 10:59:48.000000000 -0400
+++ xf86-video-i810-1.6.5/src/i830_driver.c	2009-05-14 11:03:43.000000000 -0400
@@ -3865,6 +3865,13 @@ I830BIOSPreInit(ScrnInfoPtr pScrn, int f
    pI830->PciTag = pciTag(pI830->PciInfo->bus, pI830->PciInfo->device,
 			  pI830->PciInfo->func);
 
+   if (IS_I9XX(pI830)) {
+       xf86DrvMsg(pScrn->scrnIndex, X_WARNING,
+          "Running on a post-865 chip, this is not a tested configuration\n");
+       xf86DrvMsg(pScrn->scrnIndex, X_WARNING,
+	  "The 'intel' driver is the supported driver for this chipset\n");
+   }
+
     /* Allocate an entity private if necessary */
     if (xf86IsEntityShared(pScrn->entityList[0])) {
 	pI830Ent = xf86GetEntityPrivate(pScrn->entityList[0],
